subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -51,56 +74,76 @@ public HierarchicalFileWatcherUpdater(FileWatcher watcher) {
+        updateWatchedDirectories();
+        watchedRootProjectDirectoriesFromPreviousBuild.removeIf(
+            rootProjectDirectory -> !watchedHierarchies.contains(rootProjectDirectory.path)
+        );
+    }
+
+    @Override
+    public void updateRootProjectDirectories(Collection<File> updatedRootProjectDirectories) {
+        Set<Path> rootPaths = updatedRootProjectDirectories.stream()
             .map(File::toPath)
             .map(Path::toAbsolutePath)
             .collect(Collectors.toSet());
-        projectRootDirectories.clear();
-        for (Path rootPathToWatch : WatchRootUtil.resolveRootsToWatch(rootPaths)) {
-            projectRootDirectories.put(rootPathToWatch, rootPathToWatch.toString() + File.separator);
-        }
-        LOGGER.info("Now considering {} as root directories to watch", projectRootDirectories.keySet());
+        Set<Path> newRootProjectDirectories = WatchRootUtil.resolveRootsToWatch(rootPaths);
+        LOGGER.info("Now considering {} as root directories to watch", newRootProjectDirectories);
+
+        knownRootProjectDirectoriesFromCurrentBuild.clear();
+        newRootProjectDirectories.stream()
+            .map(RootProjectDirectory::new)
+            .forEach(knownRootProjectDirectoriesFromCurrentBuild::add);
+        watchedRootProjectDirectoriesFromPreviousBuild.removeAll(knownRootProjectDirectoriesFromCurrentBuild);
+
         updateWatchedDirectories();
     }
 
     private void updateWatchedDirectories() {
         Set<Path> directoriesToWatch = new HashSet<>();
-        shouldWatchDirectories.elementSet().forEach(shouldWatchDirectory -> {
+        trackedDirectoriesForSnapshot.values().forEach(shouldWatchDirectory -> {
             String shouldWatchDirectoryPathString = shouldWatchDirectory.toString();
-            for (Map.Entry<Path, String> entry : projectRootDirectories.entrySet()) {
-                Path projectRootDirectory = entry.getKey();
-                String projectRootDirectoryPrefix = entry.getValue();
-                if (shouldWatchDirectoryPathString.startsWith(projectRootDirectoryPrefix)) {
-                    directoriesToWatch.add(projectRootDirectory);
-                    return;
-                }
+            if (maybeWatchRootProjectDirectory(directoriesToWatch, shouldWatchDirectoryPathString, knownRootProjectDirectoriesFromCurrentBuild)) {
+                return;
+            }
+            if (maybeWatchRootProjectDirectory(directoriesToWatch, shouldWatchDirectoryPathString, watchedRootProjectDirectoriesFromPreviousBuild)) {
+                return;
             }
             directoriesToWatch.add(shouldWatchDirectory);
         });
 
         updateWatchedDirectories(WatchRootUtil.resolveRootsToWatch(directoriesToWatch));
     }
 
+    private static boolean maybeWatchRootProjectDirectory(Set<Path> directoriesToWatch, String shouldWatchDirectoryPathString, Set<RootProjectDirectory> rootProjectDirectories) {
```

Comment:
I have some issues with this method. It does not actually watch directories, it enables the watching of the first root project directory that contains the given path. So I'd like to change the name of this. I think we can do better than using `maybe`, too, by giving some more context on when the action happens and when it doesn't.

Another is that the parameters are hard to read. Which are inputs and which are mutated by the method? Neither the names nor the order gives much insight into that.

Yet another issue is the mutation itself: we mutate `directoriesToWatch` in both this method and also in the caller method.

I think it would be better to unroll this method, maybe like this:

```java
    private void updateWatchedDirectories() {
        Set<Path> directoriesToWatch = trackedDirectoriesForSnapshot.values().stream()
            .map(shouldWatchDirectory -> Stream.concat(knownRootProjectDirectoriesFromCurrentBuild.stream(), watchedRootProjectDirectoriesFromPreviousBuild.stream())
                .filter(rootProjectDir -> rootProjectDir.isAncestorOf(shouldWatchDirectory))
                .findAny()
                .map(RootProjectDirectory::getPath)
                .orElse(shouldWatchDirectory))
            .collect(Collectors.toSet());

        updateWatchedDirectories(WatchRootUtil.resolveRootsToWatch(directoriesToWatch));
    }
```

We could go further perhaps and have `WatchRootUtil.resolveRootsToWatch()` take a stream, too.